File: internal/api/clientgrpc/subscriptions.go
package clientgrpc

import (
	"context"
	"encoding/base64"
	"time"

	clientv1 "github.com/torchwooddev/torchwood/genproto/client/v1"
	sharedv1 "github.com/torchwooddev/torchwood/genproto/shared/v1"
	appsubs "github.com/torchwooddev/torchwood/internal/app/subscriptions"
	domainsubs "github.com/torchwooddev/torchwood/internal/domain/subscriptions"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// SubscriptionsService 是终端用户订阅面 gRPC handler（薄：DTO 映射）。
type SubscriptionsService struct {
	clientv1.UnimplementedSubscriptionsServiceServer
	subs *appsubs.Subscriptions
}

// NewSubscriptionsService constructs the client subscriptions service.
func NewSubscriptionsService(subs *appsubs.Subscriptions) *SubscriptionsService {
	return &SubscriptionsService{subs: subs}
}

func (s *SubscriptionsService) ListPlans(ctx context.Context, req *clientv1.ListPlansRequest) (*clientv1.ListPlansResponse, error) {
	before, err := decodeSubCursor(req.GetPageToken())
	if err != nil {
		return nil, status.Error(codes.InvalidArgument, "invalid page token")
	}
	plans, err := s.subs.ListClientPlans(ctx, int(req.GetPageSize()), before)
	if err != nil {
		return nil, err
	}
	out := make([]*clientv1.SubscriptionPlan, len(plans))
	for i := range plans {
		out[i] = mapClientPlan(&plans[i])
	}
	meta := &sharedv1.ListResponseMeta{PageSize: req.GetPageSize()}
	if len(plans) > 0 {
		meta.NextPageToken = encodeSubCursor(plans[len(plans)-1].CreatedAt)
	}
	return &clientv1.ListPlansResponse{Plans: out, Meta: meta}, nil
}

func (s *SubscriptionsService) Subscribe(ctx context.Context, req *clientv1.SubscribeRequest) (*clientv1.SubscribeResponse, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "invalid request")
	}
	result, err := s.subs.Subscribe(ctx, appsubs.SubscribeCommand{
		PlanCode:         req.GetPlanCode(),
		Mode:             domainsubs.Mode(req.GetMode()),
		IdempotencyKey:   req.GetIdempotencyKey(),
		BillingAssetCode: req.GetBillingAssetCode(),
	})
	if err != nil {
		return nil, err
	}
	mapped := mapClientSub(result.Subscription, result.Plan)
	mapped.PaymentUrl = result.PaymentURL
	return &clientv1.SubscribeResponse{
		Subscription:     mapped,
		IdempotentReplay: result.IdempotentReplay,
		PaymentUrl:       result.PaymentURL,
		OrderId:          result.OrderID,
	}, nil
}

func (s *SubscriptionsService) GetMySubscription(ctx context.Context, req *clientv1.GetMySubscriptionRequest) (*clientv1.Subscription, error) {
	var planCode string
	if req != nil {
		planCode = req.GetPlanCode()
	}
	sub, plan, err := s.subs.GetMySubscription(ctx, planCode)
	if err != nil {
		return nil, err
	}
	return mapClientSub(sub, plan), nil
}

func (s *SubscriptionsService) Cancel(ctx context.Context, req *clientv1.CancelRequest) (*clientv1.Subscription, error) {
	if req == nil || req.GetSubscriptionId() == "" {
		return nil, status.Error(codes.InvalidArgument, "subscription_id is required")
	}
	sub, plan, err := s.subs.CancelAtPeriodEnd(ctx, req.GetSubscriptionId())
	if err != nil {
		return nil, err
	}
	return mapClientSub(sub, plan), nil
}

func encodeSubCursor(t time.Time) string {
	return base64.RawURLEncoding.EncodeToString([]byte(t.UTC().Format(time.RFC3339Nano)))
}

func decodeSubCursor(token string) (time.Time, error) {
	if token == "" {
		return time.Time{}, nil
	}
	raw, err := base64.RawURLEncoding.DecodeString(token)
	if err != nil {
		return time.Time{}, err
	}
	return time.Parse(time.RFC3339Nano, string(raw))
}

func mapClientPlan(p *domainsubs.Plan) *clientv1.SubscriptionPlan {
	if p == nil {
		return nil
	}
	return &clientv1.SubscriptionPlan{
		Id:            p.ID,
		Code:          p.Code,
		Name:          p.Name,
		Amount:        p.Amount,
		Currency:      p.Currency,
		Interval:      string(p.Interval),
		IntervalDays:  p.IntervalDays,
		GraceDays:     p.GraceDays,
		TrialDays:     p.TrialDays,
		Benefits:      mapClientBenefits(p.Benefits),
	}
}

func mapClientSub(sub *domainsubs.Subscription, plan *domainsubs.Plan) *clientv1.Subscription {
	if sub == nil {
		return nil
	}
	out := &clientv1.Subscription{
		Id:                  sub.ID,
		PlanId:              sub.PlanID,
		Mode:                string(sub.Mode),
		Status:              string(sub.Status),
		CurrentPeriodStart:  timestamppb.New(sub.CurrentPeriodStart),
		CurrentPeriodEnd:    timestamppb.New(sub.CurrentPeriodEnd),
		CancelAtPeriodEnd:   sub.CancelAtPeriodEnd,
		Benefits:            mapClientBenefits(sub.Benefits),
		CreatedAt:           timestamppb.New(sub.CreatedAt),
	}
	if plan != nil {
		out.PlanCode = plan.Code
	}
	if sub.GraceUntil != nil {
		out.GraceUntil = timestamppb.New(*sub.GraceUntil)
	}
	return out
}

func mapClientBenefits(b domainsubs.Benefits) *clientv1.Benefits {
	out := &clientv1.Benefits{}
	for _, g := range b.Grants {
		item := &clientv1.BenefitGrant{AssetCode: g.AssetCode, Quantity: g.Quantity}
		if g.ExpiresIn != nil {
			item.ExpiresIn = g.ExpiresIn
		}
		out.Grants = append(out.Grants, item)
	}
	for _, e := range b.Entitlements {
		out.Entitlements = append(out.Entitlements, &clientv1.BenefitEntitlement{
			AssetCode: e.AssetCode,
			Tier:      e.Tier,
		})
	}
	return out
}

File: console/src/api/subscriptions.ts
import { api } from "./client";
import type { ApiRequestConfig } from "./client";

export interface SubscriptionPlan {
  id: string;
  project_id?: string;
  code: string;
  name: string;
  amount: string;
  currency: string;
  interval: string;
  interval_days?: string;
  grace_days?: number;
  trial_days?: number;
  status?: string;
  created_at?: string;
  updated_at?: string;
}

export interface Subscription {
  id: string;
  project_id?: string;
  user_id?: string;
  plan_id?: string;
  plan_code?: string;
  mode: string;
  status: string;
  current_period_start?: string;
  current_period_end?: string;
  cancel_at_period_end?: boolean;
  grace_until?: string;
  created_at?: string;
}

export async function listPlans(): Promise<SubscriptionPlan[]> {
  const res = await api.get<{ plans: SubscriptionPlan[] }>("/server/subscriptions/plans");
  return res.data.plans ?? [];
}

export async function getPlan(planId: string): Promise<SubscriptionPlan> {
  const res = await api.get<SubscriptionPlan>(`/server/subscriptions/plans/${planId}`);
  return res.data;
}

export async function createPlan(input: {
  code: string;
  name: string;
  amount: string;
  currency: string;
  interval: string;
  interval_days?: string;
  grace_days?: number;
  trial_days?: number;
}): Promise<SubscriptionPlan> {
  const res = await api.post<SubscriptionPlan>("/server/subscriptions/plans", input);
  return res.data;
}

export async function updatePlan(
  planId: string,
  input: { name?: string; status?: string }
): Promise<SubscriptionPlan> {
  const res = await api.patch<SubscriptionPlan>(`/server/subscriptions/plans/${planId}`, {
    plan_id: planId,
    ...input,
  });
  return res.data;
}

export async function deletePlan(planId: string, config?: ApiRequestConfig): Promise<void> {
  await api.delete(`/server/subscriptions/plans/${planId}`, config);
}

export async function listSubscriptions(): Promise<Subscription[]> {
  const res = await api.get<{ subscriptions: Subscription[] }>("/server/subscriptions");
  return res.data.subscriptions ?? [];
}

export async function getSubscription(subscriptionId: string): Promise<Subscription> {
  const res = await api.get<Subscription>(`/server/subscriptions/${subscriptionId}`);
  return res.data;
}

export async function cancelSubscription(subscriptionId: string, reason?: string): Promise<Subscription> {
  const res = await api.post<Subscription>(`/server/subscriptions/${subscriptionId}:cancel`, {
    subscription_id: subscriptionId,
    reason,
  });
  return res.data;
}

export async function expireSubscription(subscriptionId: string, reason?: string): Promise<Subscription> {
  const res = await api.post<Subscription>(`/server/subscriptions/${subscriptionId}:expire`, {
    subscription_id: subscriptionId,
    reason,
  });
  return res.data;
}
